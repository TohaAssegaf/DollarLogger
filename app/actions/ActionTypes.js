import {
    Payment
} from '/app/store/state/PaymentState'

// Set budget total.
export const SET_BUDGET_TOTAL_REQUEST = 'SET_BUDGET_TOTAL_REQUEST'
export type SetBudgetTotalRequestAction = {
    type: SET_BUDGET_TOTAL_REQUEST
}
export const SET_BUDGET_TOTAL_SUCCESS = 'SET_BUDGET_TOTAL_SUCCESS'
export type SetBudgetTotalSuccessAction = {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const SET_BUDGET_TOTAL_FAILURE = 'SET_BUDGET_TOTAL_FAILURE'
export type SetBudgetTotalFailureAction = {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Get budget total.
export const GET_BUDGET_TOTAL_REQUEST = 'GET_BUDGET_TOTAL_REQUEST'
export type GetBudgetTotalRequestAction = {
    type: GET_BUDGET_TOTAL_REQUEST
}
export const GET_BUDGET_TOTAL_SUCCESS = 'GET_BUDGET_TOTAL_SUCCESS'
export type GetBudgetTotalSuccessAction = {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const GET_BUDGET_TOTAL_FAILURE = 'GET_BUDGET_TOTAL_FAILURE'
export type GetBudgetTotalFailureAction = {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Create payment.
export const CREATE_PAYMENT_REQUEST = 'CREATE_PAYMENT_REQUEST'
export type CreatePaymentRequestAction = {
    type: CREATE_PAYMENT_REQUEST
}
export const CREATE_PAYMENT_SUCCESS = 'CREATE_PAYMENT_SUCCESS'
export type CreatePaymentSuccessAction = {
    type: CREATE_PAYMENT_SUCCESS,
    payment: Payment
}
export const CREATE_PAYMENT_FAILURE = 'CREATE_PAYMENT_FAILURE'
export type CreatePaymentFailureAction = {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage: string
}

export type Action = |
    SetBudgetTotalRequestAction |
    SetBudgetTotalSuccessAction |
    SetBudgetTotalFailureAction |
    GetBudgetTotalRequestAction |
    GetBudgetTotalSuccessAction |
    GetBudgetTotalFailureAction |
    CreatePaymentRequestAction |
    CreatePaymentSuccessAction |
    CreatePaymentFailureAction