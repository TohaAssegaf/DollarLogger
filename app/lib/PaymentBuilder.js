import * as DateUtils from './DateUtils'

const ONE_DAY = 86400000
const ONE_WEEK = ONE_DAY * 7

export default class PaymentBuilder {
  id: number
  total: number
  name: string
  date: Date
  splitCount: number

  constructor(payment: Payment) {
    if (payment === undefined) {
      this.splitCount = 1
    } else {
      this.id = payment.id
      this.total = payment.total
      this.name = payment.name
      this.date = payment.date
      this.splitCount = payment.paymentContributions.length
    }
  }

  setId(id: number): PaymentBuilder {
    this.id = id
    return this
  }

  setTotal(total: number): PaymentBuilder {
    this.total = total
    return this
  }

  setName(name: string): PaymentBuilder {
    this.name = name
    return this
  }

  setDate(date: Date): PaymentBuilder {
    this.date = date
    return this
  }

  setSplitCount(splitCount: number): PaymentBuilder {
    this.splitCount = splitCount
    return this
  }

  build(): Payment {
    this.reservePaymentId()
    const paymentContributions = this.buildPaymentContributions()
    return {
      id: this.id,
      total: this.total,
      name: this.name,
      date: this.date,
      paymentContributions,
    }
  }

  reservePaymentId() {
    // TODO(renzobautista): Separate ID generation into a new class so it can be mocked
    if (!this.id) {
      this.id = Date.now()
    }
  }

  buildPaymentContributions(): Array<PaymentContribution> {
    const totals = this.getPaymentContributionTotals()
    const dates = this.getPaymentContributionDates()
    return totals.map((total, i) => this.buildPaymentContribution(total, dates[i], i))
  }

  getPaymentContributionTotals(): Array<number> {
    const flooredValue = Math.floor(this.total / this.splitCount)
    // How many pennies are left when multiplying flooredValue by splitCount.
    const remainder = this.total % (flooredValue * this.splitCount)
    let totals = []
    for (let i = 0; i < this.splitCount; i++) {
      // Evenly split the remaining cents across the first remainder contributions.
      const spareCent = i < remainder ? 1 : 0
      totals.push(flooredValue + spareCent)
    }
    return totals
  }

  getPaymentContributionDates(): Array<Date> {
    // First date is always the provided date.
    let dates = [this.date]
    // Every subsequent date is the Monday of the successive weeks.
    for (let i = 1; i < this.splitCount; i++) {
      const dateIncrement = ONE_WEEK * i
      dates.push(DateUtils.getLastMonday(new Date(this.date.getTime() + dateIncrement)))
    }
    return dates
  }

  buildPaymentContribution(total: number, date: date, i: index): PaymentContribution {
    return {
      displayName: this.buildDisplayName(i),
      total,
      date,
      paymentId: this.id,
    }
  }

  buildDisplayName(i: number): string {
    // If only 1, then use same name. Otherwise, add a suffix showing installment number.
    // Installment number needs to be incremented by 1 because it is 0-indexed.
    return this.splitCount === 1 ? this.name : this.name + ` (${i + 1}/${this.splitCount})`
  }
}
