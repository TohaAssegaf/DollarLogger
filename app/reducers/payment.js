import {
  GET_PAYMENTS_REQUEST,
  GET_PAYMENTS_SUCCESS,
  GET_PAYMENTS_FAILURE,
  Action
} from '~/app/actions/ActionTypes'
import {
  Payment,
  PaymentState
} from '~/app/store/state/PaymentState'

const initialState: PaymentState = {
  payments: [],
  isLoading: true,
  errorMessage: '',
}

export default function (state: PaymentState = initialState, action: Action): PaymentState {
  switch (action.type) {
  case GET_PAYMENTS_REQUEST:
    return getPaymentsRequest(state, action)
  case GET_PAYMENTS_SUCCESS:
    return getPaymentsSuccess(state, action)
  case GET_PAYMENTS_FAILURE:
    return getPaymentsFailure(state, action)
  default:
    return state
  }
}

function getPaymentsRequest(
  state: PaymentState, action: GetPaymentsRequestAction): PaymentState {
  return Object.assign({}, state, {
    isLoading: true,
    errorMessage: ""
  })
}

function getPaymentsSuccess(
  state: PaymentState, action: GetPaymentsSuccessAction): PaymentState {
  let payments = action.payments.slice()
  return Object.assign({}, state, {
    payments,
    isLoading: false,
  })
}

function getPaymentsFailure(
  state: PaymentState, action: GetPaymentsFailureAction): PaymentState {
  return Object.assign({}, state, {
    errorMessage: action.errorMessage,
    isLoading: false,
  })
}
