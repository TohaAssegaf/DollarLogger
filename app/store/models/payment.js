import { PAYMENTS_ASYNC_STORAGE_KEY } from '~/app/config/storage'
import { AsyncStorage } from 'react-native'

export function getPayments() {
  return AsyncStorage.getItem(PAYMENTS_ASYNC_STORAGE_KEY)
    .then(payments => {
      if (!payments) {
        return []
      }
      return JSON.parse(payments).map(payment => parsePayment(payment))
    })
}

/** Need to parse all date fields into actual Date objects instead of numbers. */
function parsePayment(payment) {
  let paymentContributions = payment.paymentContributions.map(
    paymentContribution =>
      Object.assign({}, paymentContribution, { date: new Date(paymentContribution.date) }))
  return Object.assign({}, payment, { date: new Date(payment.date), paymentContributions })
}

export function addPayment(total: number, name: string, date: Date) {
  return getPayments().then(payments => {
    // TODO(renzobautista): Separate ID generation into a new class so it can be mocked
    const id = new Date().getTime()
    const payment = {
      id,
      total,
      name,
      date,
      paymentContributions: [
        {
          total,
          date,
        }
      ],
    }
    payments.push(payment)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(payments))
      .then(() => payments)
  })
}

export function updatePayment(payment: Payment) {
  return getPayments().then(payments => {
    const updatedPayments = payments.map(
      storedPayment => storedPayment.id === payment.id ? payment : storedPayment)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(updatedPayments))
      .then(() => updatedPayments)
  })
}

export function deletePayment(id: number) {
  return getPayments().then(payments => {
    const updatedPayments = payments.filter(payment => payment.id != id)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(updatedPayments))
      .then(() => updatedPayments)
  })
}
