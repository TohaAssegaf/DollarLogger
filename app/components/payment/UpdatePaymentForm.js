import BasePaymentForm from './BasePaymentForm'
import actions from '~/app/actions'
import PaymentBuilder from '~/app/lib/PaymentBuilder'
import React from 'react'
import { connect } from 'react-redux'

class UpdatePaymentForm extends React.Component {
  updatePayment(total, name, date, splitCount) {
    const payment =
      new PaymentBuilder(this.props.payment)
        .setTotal(total)
        .setName(name)
        .setDate(date)
        .setSplitCount(splitCount)
        .build()
    this.props.updatePayment(payment)

    // Temporary hack. This should check state with a componentDidUpdate for write to be complete.
    this.props.navigation.goBack()
  }

  render() {
    return (
      <BasePaymentForm
        payment={this.props.payment}
        onSubmit={
          (total, name, date, splitCount) => this.updatePayment(total, name, date, splitCount)}
      />
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    updatePayment: payment => {
      dispatch(actions.updatePayment(payment))
    }
  }
}

export default connect((state) => ({}), mapDispatchToProps)(UpdatePaymentForm)
