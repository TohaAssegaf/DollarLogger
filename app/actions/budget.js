import {
  SET_BUDGET_TOTAL_REQUEST,
  SET_BUDGET_TOTAL_SUCCESS,
  SET_BUDGET_TOTAL_FAILURE,
  GET_BUDGET_TOTAL_REQUEST,
  GET_BUDGET_TOTAL_SUCCESS,
  GET_BUDGET_TOTAL_FAILURE,
  SetBudgetTotalRequestAction,
  SetBudgetTotalSuccessAction,
  SetBudgetTotalFailureAction,
  GetBudgetTotalRequestAction,
  GetBudgetTotalSuccessAction,
  GetBudgetTotalFailureAction
} from './ActionTypes'
import * as BudgetModel from '/app/store/models/budget'

export function setBudgetTotalRequest(): SetBudgetTotalRequestAction {
  return {
    type: SET_BUDGET_TOTAL_REQUEST
  }
}

export function setBudgetTotalSuccess(total: number): SetBudgetTotalSuccessAction {
  return {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function setBudgetTotalFailure(errorMessage: string): SetBudgetTotalFailureAction {
  return {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function setBudgetTotal(total: number) {
  return function(dispatch) {
    if (total <= 0) {
      dispatch(setBudgetTotalFailure("Budget must be greater than 0"))
    }
    if (!Number.isInteger(total)) {
      dispatch(setBudgetTotalFailure("Budget must be an integer"))
    }
    dispatch(setBudgetTotalRequest(total))
    return BudgetModel.setTotal(total)
      .then(() => {
        dispatch(setBudgetTotalSuccess(total))
        dispatch(NavigationActions.back())
      })
      .catch(error => dispatch(setBudgetTotalFailure(error.message)))
  }
}

export function getBudgetTotalRequest(): GetBudgetTotalRequestAction {
  return {
    type: GET_BUDGET_TOTAL_REQUEST
  }
}

export function getBudgetTotalSuccess(total: ?number): GetBudgetTotalSuccessAction {
  return {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function getBudgetTotalFailure(errorMessage: string): GetBudgetTotalFailureAction {
  return {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function getBudgetTotal() {
  return function(dispatch) {
    dispatch(getBudgetTotalRequest())
    return BudgetModel.getTotal()
      .then(total => {
        dispatch(getBudgetTotalSuccess(total))
      })
      .catch(error => dispatch(getBudgetTotalFailure(error.message)))
  }
}
