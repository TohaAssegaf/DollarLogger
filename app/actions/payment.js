import {
  CREATE_PAYMENT_REQUEST,
  CREATE_PAYMENT_SUCCESS,
  CREATE_PAYMENT_FAILURE,
  CreatePaymentRequestAction,
  CreatePaymentSuccessAction,
  CreatePaymentFailureAction
} from './ActionTypes'
import * as PaymentModel from '/app/store/models/payment'
import {
  Payment
} from '/app/store/state/PaymentState'

export function createPaymentRequest(): CreatePaymentRequestAction {
  return {
    type: CREATE_PAYMENT_REQUEST
  }
}

export function createPaymentSuccess(payment: Payment): CreatePaymentSuccessAction {
  return {
    type: CREATE_PAYMENT_SUCCESS,
    payment
  }
}

export function createPaymentFailure(errorMessage: string): CreatePaymentFailureAction {
  return {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage
  }
}

export function createPayment(total: number, name: string, date: Date) {
  return function (dispatch) {
    if (!Number.isInteger(total)) {
      dispatch(createPaymentFailure("Payment total must be an integer"))
    }
    dispatch(createPaymentRequest())
    return PaymentModel.createPayment(total, name, date)
      .then(payment => dispatch(createPaymentSuccess(payment)))
      .catch(error => dispatch(createPaymentFailure(error.message)))
  }
}