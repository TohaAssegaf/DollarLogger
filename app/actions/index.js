import * as budget from './budget'

export default Object.assign({}, budget)
