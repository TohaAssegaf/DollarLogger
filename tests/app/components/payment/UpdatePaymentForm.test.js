import actions from '~/app/actions'
import UpdatePaymentForm from '~/app/components/payment/UpdatePaymentForm'
import PaymentBuilder from '~/app/lib/PaymentBuilder'
import * as PaymentModel from '~/app/store/models/payment'
import enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import MockDate from 'mockdate'
import React from 'react'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'

enzyme.configure({ adapter: new Adapter() })
const navigation = { goBack: jest.fn() }
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

let mockPayments = []
jest.mock('../../../../app/store/models/payment', () => {
  return {
    updatePayment: jest.fn((item, payment) => {
      return new Promise((resolve, reject) => {
        mockPayments = mockPayments.map(
          mockPayment => mockPayment.id === payment.id ? payment : mockPayment)
        resolve(mockPayments)
      })
    })
  }
})

it('renders correctly', () => {
  const payment = new PaymentBuilder()
    .setId(1)
    .setTotal(1000)
    .setName("Test payment")
    .setDate(new Date(2018, 4, 3))
    .build()
  const store = mockStore({ payment: { payments: [payment] } })
  const rendered = renderer.create(<UpdatePaymentForm
    payment={payment}
    store={store}
    navigation={navigation}
  />).toJSON();
  expect(rendered).toMatchSnapshot();
});

it('dispatches update payment action and navigates back', () => {
  const payment =
    new PaymentBuilder()
      .setId(1)
      .setTotal(1000)
      .setName("Test payment")
      .setDate(new Date(2018, 4, 2))
      .setSplitCount(4)
      .build()
  const newTotal = 20000
  const newName = "New name"
  const newDate = new Date(2018, 4, 3)
  const newSplitCount = 1
  const updatedPayment =
    new PaymentBuilder()
      .setId(1)
      .setTotal(newTotal)
      .setName(newName)
      .setDate(newDate)
      .setSplitCount(newSplitCount)
      .build()
  const store = mockStore({ payment: { payments: [payment] } })
  const wrapper = shallow(
    <UpdatePaymentForm payment={payment} store={store} navigation={navigation} />)
    .dive({ context: { store } })

  wrapper.simulate('submit', newTotal, newName, newDate, newSplitCount)

  expect(store.getActions()).toContainEqual(actions.getPaymentsRequest())
  expect(PaymentModel.updatePayment.mock.calls).toEqual([[updatedPayment]])
  expect(navigation.goBack.mock.calls).toHaveLength(1)
});
