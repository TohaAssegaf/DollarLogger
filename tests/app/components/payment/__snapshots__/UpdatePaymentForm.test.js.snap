// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePaymentForm renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 20,
            },
            Object {
              "fontSize": 25,
            },
          ]
        }
      >
        $
      </Text>
      <TextInput
        allowFontScaling={true}
        keyboardType="numeric"
        onBlur={[Function]}
        onChangeText={[Function]}
        style={
          Array [
            Object {
              "fontSize": 20,
              "minWidth": 100,
            },
            Object {
              "flex": 1,
              "fontSize": 25,
            },
          ]
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value="10.00"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Name"
      style={
        Object {
          "fontSize": 25,
        }
      }
      underlineColorAndroid="rgba(0,0,0,0)"
      value="Test payment"
    />
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "width": 142,
          },
          Object {
            "width": "auto",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
            },
            Object {
              "alignItems": "flex-start",
              "height": "auto",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "#aaa",
                "borderWidth": 1,
                "flex": 1,
                "height": 40,
                "justifyContent": "center",
              },
              Object {
                "alignItems": "flex-start",
                "borderWidth": 0,
                "height": "auto",
              },
              false,
              false,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#333",
                },
                Object {
                  "fontSize": 25,
                },
              ]
            }
          >
            05/03/2018
          </Text>
        </View>
        <Modal
          animationType="none"
          hardwareAccelerated={false}
          onRequestClose={[Function]}
          supportedOrientations={
            Array [
              "portrait",
              "portrait-upside-down",
              "landscape",
              "landscape-left",
              "landscape-right",
            ]
          }
          transparent={true}
          visible={false}
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "flex-end",
                  "backgroundColor": "#00000077",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "height": 0,
                      "overflow": "hidden",
                    }
                  }
                >
                  <View
                    pointerEvents="auto"
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "borderTopColor": "#ccc",
                            "borderTopWidth": 1,
                            "marginTop": 42,
                          },
                          undefined,
                        ]
                      }
                    >
                      <RCTDatePicker
                        date={1525320000000}
                        mode="date"
                        onChange={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "height": 216,
                          }
                        }
                      />
                    </View>
                  </View>
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "height": 42,
                          "justifyContent": "center",
                          "padding": 20,
                          "position": "absolute",
                          "top": 0,
                        },
                        Object {
                          "left": 0,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#46cf98",
                            "fontSize": 16,
                          },
                          Object {
                            "color": "#666",
                          },
                          undefined,
                        ]
                      }
                    >
                      CANCEL
                    </Text>
                  </View>
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "height": 42,
                          "justifyContent": "center",
                          "padding": 20,
                          "position": "absolute",
                          "top": 0,
                        },
                        Object {
                          "right": 0,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#46cf98",
                            "fontSize": 16,
                          },
                          undefined,
                        ]
                      }
                    >
                      OK
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </Modal>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#222222",
        "borderBottomWidth": 0.5,
        "height": 70,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="Weekly installments (1 by default)"
      style={
        Object {
          "fontSize": 25,
        }
      }
      underlineColorAndroid="rgba(0,0,0,0)"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginHorizontal": 20,
        "marginTop": 20,
      }
    }
  >
    <View
      accessibilityLabel="Save payment"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#6A1B9A",
              },
            ]
          }
        >
          SAVE
        </Text>
      </View>
    </View>
  </View>
</View>
`;
